docs/rfcs/rfc-040-relay-first-api.md

```diff
@@ -0,0 +1,1351 @@
+
+```graphql
+query PaginatedMovies {
+    moviesConnection(first: 10, after: "my-cursor", sort: { edges: { node: { title: DESC } } }) {
+        edges {
+            node {
+                title
+            }
+            cursor
+        }
+        pageInfo {
+            startCursor
+            hasPreviousPage
+            hasNextPage
+            endCursor
+        }
+    }
+}
+```
+
+Following Relay spec, every `*Connection` type has the `pageInfo` property, and each `edge` has a cursor.
+
+#### Offset pagination
+
+Offset pagination could be enabled in configuration, in this case the properties `limit` and `offset` would be used. Page info and cursor would not be returned in this case.
+
+```graphql
+query PaginatedMovies {
+    moviesConnection(limit: 10, offset: 5, sort: { edges: { node: { title: DESC } } }) {
+        edges {
+            node {
+                title
+            }
+            cursor
+        }
+    }
+}
+```
+
+Note that if offset-based pagination is enabled instead of cursor based, the API will not be Relay compliant.
+
+#### Back pagination
+
+Following Relay spec, the fields `last` and `before` can be used instead of `first` and `after` to do backwards pagination, this could be toggled in the server configuration.
+
+### Aggregation
+
+Aggregations have been moved to be part of the projection on connection queries:
+
+```graphql
+query ShortestMovieTitleAndCount {
+    moviesConnection {
+        aggregation {
+            nodes {
+                title {
+                    shortest
+                }
+                count
+            }
+        }
+    }
+}
+```
+
+This means that a projection can be returned along with the edges. Nested aggregations can be queried in the same fashion. `*Aggregation` types returned in a relationship will have the `node` and `fields` properties with the respective aggregation:
+
+```graphql
+query AggregateActorsPerMovie {
+    moviesConnection {
+        edges {
+            node {
+                actors {
+                    aggregation {
+                        nodes {
+                            count
+                        }
+                        edges {
+                            fields {
+                                year {
+                                    min
+                                    max
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+> Aggregations are not available for 1-1 relationships
+
+### Fulltext
+
+If fulltext is enable for an entity using the `@fulltext` directive, the `fulltext` input will be available, as well as the `score` fields and filters as a `Float`:
+
+```graphql
+query Query {
+    moviesConnection(
+        fulltext: { phrase: "Test", index: MovieTitle }
+        where: { edges: { fulltext: { score: { gt: 5 } } } }
```

Comment:
Is there a reason that the score filter cannot only be included in the top-level fulltext argument? It feels weird to be to be able to filter by a score that may not exist and I can see this leading to confusion/bugs if we are not careful.

I was thinking the API for this should be something like this:

```graphql
query Query {
    moviesConnection(
        fulltext: { phrase: "Test", index: MovieTitle, where: { edges: { fulltext: { score: { gt: 5 } } } } }
    ) {
        totalCount
        edges {
            fulltext {
                score
            }
        }
    }
}
```

We could then only allow the `score` filter to be used within this `where` input. To me that feels similar to how we do it for `create`/`connect`/`update` but does come with the down side of some extra type definitions.

Reply:
The issue here is that you will lose the ability to combine the FullText predicate with the common one.

Reply:
I was suggesting having access to all the filters at this level e.g. 

```graphql
query Query {
    moviesConnection(
        fulltext: { phrase: "Test", index: MovieTitle, where: { edges: { node: { title: { equals: "foo" } }, fulltext: { score: { gt: 5 } } } } }
    ) {
        totalCount
        edges {
            fulltext {
                score
            }
        }
    }
}
```

Reply:
You would have still to maintain the other `where` input, imagine that you want to select a Movie filtered by the FullText score combined with a very long and complex predicate. Then, after some development, you decide that you don't want any more filtering by the FullText score. Now, the only options will be:
 - Keep the predicates inside the fulltext input without using any fulltext(assuming that we support it).
- Migrate the whole predicate to the normal where input (the very likely scenario). 
I see your concerns where are originated from, but in my opinion, the alternatives are worse!

Reply:
I think the proposal here is more confusing; there should be a single `where` which allows you to do all combinations of filtering. However, I totally agree on the confusion of splitting the "fulltext logic", but this is a trade-off of having it all accessible in one field.